src/inmanta/agent/agent.py

```diff
@@ -1332,52 +1490,81 @@ async def on_disconnect(self) -> None:
-                                    hash_value=source["hash"],
-                                    _client=sync_client,
-                                )
+                    sync_client = SyncClient(client=self._client, ioloop=self._io_loop)
+                    LOGGER.debug("Installing handler %s version=%d", resource_type, version)
+                    requirements: set[str] = set()
+                    sources: list["ModuleSource"] = []
+                    for source in result.result["data"]:
+                        sources.append(
+                            ModuleSource(
+                                name=source["module_name"],
+                                is_byte_code=source["is_byte_code"],
+                                hash_value=source["hash"],
+                                _client=sync_client,
                             )
-                            requirements.update(source["requirements"])
+                        )
+                        requirements.update(source["requirements"])
 
-                        if pip_config is None:
-                            pip_config = await self._get_pip_config(environment, version)
-                        await self._install(sources, list(requirements), pip_config=pip_config)
-                        LOGGER.debug("Installed handler %s version=%d", rt, version)
-                        self._last_loaded[rt] = version
-                    except Exception:
-                        LOGGER.exception("Failed to install handler %s version=%d", rt, version)
-                        failed_to_load.add(rt)
+                    if pip_config is None:
+                        # TODO: what if this fails?
+                        pip_config = await self._get_pip_config(environment, version)
+
+                    resource_install_specs.append(
+                        ResourceInstallSpec(resource_type, version, pip_config, list(requirements), sources)
+                    )
+                else:
+                    # TODO what if server call fails ?
+                    pass
+
+        return sorted(resource_install_specs, key=lambda resource_install_spec: resource_install_spec.resource_type)
+
+    async def ensure_code(self, code: Sequence[ResourceInstallSpec]) -> Set[str]:
+        """Ensure that the code for the given environment and version is loaded"""
+        failed_to_load: set[str] = []
+        for resource_install_spec in code:
+            # only one logical thread can load a particular resource type at any time
+            async with self._resource_loader_lock.get(resource_install_spec.resource_type):
+                # stop if the last successful load was this one
+                # The combination of the lock and this check causes the reloads to naturally 'batch up'
+
+                # clear cache, for retry on failure
+                self._last_loaded[resource_install_spec.resource_type] = -1
```

Comment:
With the new implementation we no longer skip if it is in the cache, while I think that would still be a valid approach, right?